experimental/AlgebraicStatistics/src/MultigradedImplicitization.jl

```diff
@@ -0,0 +1,323 @@
+```
+"""
+function compute_component(mon_basis::Vector{<:MPolyDecRingElem}, phi::MPolyAnyMap)
+  image_polys = [phi(m) for m in mon_basis]
+  mons = unique!(reduce(vcat, [collect(monomials(f)) for f in image_polys]))
+  coeffs = matrix(QQ, [[coeff(f, mon) for f in image_polys] for mon in mons])
+  (r, K) = nullspace(coeffs)
+  return mon_basis * K
+end
+
+@doc raw"""
+    filter_component(deg::FinGenAbGroupElem, mon_basis::Vector{<: MPolyDecRingElem}, jac::MatElem)
+
+This function is used to skip homogeneous components which cannot contain an element of the kernel by checking if the variables in that component are independent
+in the matroid defined by the jacboian of a polynomial map. If the submatrix of the jacobian corresponding to a subset of the variables in the codomain is full rank, then
+there cannot be a relation in the kernel involving only those variables. 
+```
+"""
+function filter_component(deg::FinGenAbGroupElem, mon_basis::Vector{<: MPolyDecRingElem}, jac::MatElem)
+  # if the basis only has 1 element, then there are no generators by assumption
+  length(mon_basis) < 2 && return [true, deg]
+
+  # find the indices of all variables involved in this component
+  supp = component_support(mon_basis)
+
+  # check if the corresponding submatrix drops rank
+  # if it does not, then this component cannot contain any generator of the kernel
+  if rank(jac[:, supp]) == length(supp)
+    return [true, deg]
+  end
+
+  return [false, deg]
+end
+
+
+@doc raw"""
+    components_of_kernel(d::Int, phi::MPolyAnyMap; wp::Union{OscarWorkerPool, Nothing}=nothing)
+
+Computes all minimal generators of the kernel of the polynomial map $\phi: \mathbb{K}[x_1, \ldots, x_n] \to \mathbb{K}[t_1, \ldots, t_m]$ such that $(i, j)$ 
+with total degree at most $d$ using the main algorithm of [CH26](@cite). This implementation requires that $\phi$ be homogeneous in a multigrading which refines total
+degree which ensures the algorithm is parallelizable. The output is a dictionary whose keys are the multidegrees and values consist of a minimal generating set for that multidegree. 
+
+## Examples
+
+```jldoctest
+julia> R, x = polynomial_ring(QQ, :x => (1:2, 1:2))
+(Multivariate polynomial ring in 4 variables over QQ, QQMPolyRingElem[x[1, 1] x[1, 2]; x[2, 1] x[2, 2]])
+
+julia> S, s, t = polynomial_ring(QQ, :s => 1:2, :t => 1:2)
+(Multivariate polynomial ring in 4 variables over QQ, QQMPolyRingElem[s[1], s[2]], QQMPolyRingElem[t[1], t[2]])
+
+julia> phi = hom(R, S, [s[1]*t[1], s[1]*t[2], s[2]*t[1], s[2]*t[2]])
+Ring homomorphism
+  from multivariate polynomial ring in 4 variables over QQ
+  to multivariate polynomial ring in 4 variables over QQ
+defined by
+  x[1, 1] -> s[1]*t[1]
+  x[2, 1] -> s[1]*t[2]
+  x[1, 2] -> s[2]*t[1]
+  x[2, 2] -> s[2]*t[2]
+
+julia> components_of_kernel(2, phi)
+Dict{FinGenAbGroupElem, Vector{<:MPolyDecRingElem}} with 1 entry:
+  [1 1 1 1] => MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x[1, 1]*x[2, 2] - x[2, 1]*x[1, 2]]
```

Comment:
Mac doesn't like this line, probably because it changes the printing of Dictionnaries?
Try this?
```suggestion
julia> D = components_of_kernel(2, phi);

julia> D[1, 1, 1, 1]
MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x[1, 1]*x[2, 2] - x[2, 1]*x[1, 2]]
```
But being able to do this will require a fix for the return Dict type. See my other comment

Reply:
Or maybe this could work without the fix for the Dict?
```suggestion
julia> D = components_of_kernel(2, phi);

julia> G = parent(first(keys(D)));

julia> D[G([1, 1, 1, 1])]
MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x[1, 1]*x[2, 2] - x[2, 1]*x[1, 2]]
```

Reply:
I've overloaded getindex for these kinds of Dicts, and so in this way we get the best of both worlds.